### ISSUE TITLE
Should an error boundary prevent `console.error`?

### ISSUE THREAD

**Issue by silverwind** [facebook/react#15069]
I noticed [this unconditional `console.error`](https://github.com/facebook/react/blob/d0289c7e3a2dfc349dcce7f9eb3dee22464e97bd/packages/react-reconciler/src/ReactFiberErrorLogger.js#L86) which I'd like to prevent to keep the console clean from errors that are already "caught" in an error boundary.

Maybe a condition on `capturedError.errorBoundaryFound` could prevent this logging?

**Comment by gaearon** [comment]
For now we'd rather not allow it because it makes it too easy to swallow errors (such as if the error boundary itself is buggy).

**Comment by silverwind** [comment]
Right, I see reasoning in that, preventing error logging should require more than just the existance of a boundary.

Browsers allow to suppress logging uncaught errors to the console via `event.preventDefault()` in a `error` event handler on `window`. Maybe a something similar could be done from a error boundary too, like `error.preventDefault()`.

**Comment by anilanar** [comment]
Error swallowing should be opt-in. I'd like to swallow certain errors and not others:

```
componentDidCatch(error) {
  // I'd like to swallow the error
  if (error instanceof MyExpectedError) {
    // somehow suppress the error?
  }
}
```

**Comment by jamstooks** [comment]
I'd support this feature as it would be nice to de-clutter our test output. For now I'm just mocking `console.error` (unless it fails), but suppression of a specific error would be much better.

**Comment by EyalPerry** [comment]
@gaearon 

consider the following production use case we are facing in Wix.com's dashboard:

We are using a third party logging service, which logs all console.error automagically- which is great.
But, we want to slap on extra information special to specific Error Boundaries.
Our error boundary component receives this information as props / context,
And passes it on to sentry if an error occurs in the componentDidCatch method.
As you would imagine, we have several of these, each with it's own special information and context.

When a component fails, we have two separate errors logged in our monitoring pipeline:
One which react logs, with barely any info, which comes from console.error
And one with everything we need to know, which we explicitly logged via the third party library.

If we don't explicitly log, we lose error specific info.
We also rely on react doing the logging for us, which is a bad idea in the long run.
if we explicitly log - we get two separate events in our monitoring pipeline, which is noisy and troublesome,
Since our monitoring workflow is automated and triggers a whole lot of internal processes.

console.error is a side effect, and should not be invoked without possibility to opt out.
Not by a UI library.

That being said, I do understand where you'r coming from by saying you don't want it to be easy to swallow errors,
so how about this as middleGround: move the console.error implementation to componentDidCatch,
Which would be the default implementation.
Which, if you think of it, is the `correct` thing to do, since componentDidCatch is meant for exactly this.

you say `console.error` and I say `thirdPartyLibrary.logError`
It should be up to app developers, especially if they went through the trouble of setting up Error Boundaries.

**Comment by anilanar** [comment]
@EyalPerry hit the nail on the head. Is there anything more to say? Shall we patch `console.error`?

**Comment by tonix-tuft** [comment]
@anilanar I completely agree with you, I think we really should have the possibility to suppress the logging of some errors which have been handled in error boundaries.

I think that we should have the power to decide what to log and what not to log, especially in production.

Also, I think that, when using `create-react-app`, it would be great to have the ability to suppress the error overlay for some errors caught by the error boundary (not all, but some), but that's another story...

**Comment by ElForastero** [comment]
It would be nice to have control on logging errors to console.

For example, I like to use Error Boundaries for handling 404 exceptions in my apps. This way I don't need to call a redirect to some route or put the rendering logic inside my components.

Regularly I just throw a `NotFoundException` and seeing an error in console is not something I'm not expecting.

I would agree with @EyalPerry that moving the logging into default componentDidCatch would be a nice solution.

**Comment by anilanar** [comment]
Don’t close. We can’t seriously unconditionally log this anymore.

**Comment by art-in** [comment]
I have the opposite problem. Error is __not__ logged to console in dev build.  

`The above error occurred in the...`, but nothing is above.

So if I `console.error` it in `componentDidCatch` it works in dev, but now I have two duplicated errors logged in prod.

**Comment by zordone** [comment]
I'd like this to be fixed as well. The solution @EyalPerry suggested would be perfect.

**Comment by mariuscosta** [comment]
I vote for this too. I am also using an error handling mechanism like the one described by @ElForastero.

**Comment by kumar303** [comment]
Does anyone have a workaround for this while testing? I want to real-render a component under test, catch the error, and make an assertion without seeing error logging, especially as one [can't hide logging in jest](https://github.com/facebook/jest/issues/4156).

**Comment by MatiasManevi** [comment]
> Does anyone have a workaround for this while testing? I want to real-render a component under test, catch the error, and make an assertion without seeing error logging, especially as one [can't hide logging in jest](https://github.com/facebook/jest/issues/4156).

This is my hacky workaround:

```
const swallowErrors = yourTestFn => {
     const error = console.error
     console.error = () => {}
     yourTestFn()
     console.error = error
}

// Then your test
it('tests something without console log' , () => {
      swallowErrors(()=>{
            const wrap = mount(<ErBoundary><Child/></ErBoundary>)
            // ... expect errors ...
      })
})
```

**Comment by anilanar** [comment]
Bump.

**Comment by ZackWard** [comment]
Bump.

**Comment by Arithmetics** [comment]
bump. so many logs i want gone

**Comment by advl** [comment]
Bump

**Comment by nathggns** [comment]
This is resulting in double errors in our logging system (ErrorBoundary logs it but we also capture console.error) – extremely frustrating.

**Comment by raveclassic** [comment]
The errors are also unconditionally logged during tests where I expect components to throw and catch the errors myself with error boundaries. This is frustrating and pollutes test output.

**Comment by anilanar** [comment]
I suppose this is still not fixed in React 18 alpha. We still have hopes for stable 18 release.

**Comment by BYK** [comment]
@gaearon is there any way to prioritize this or would you accept PRs to fix it?

**Comment by ask-imran** [comment]
We are running into a similar problem as @EyalPerry  mentioned. It would be great to fix this

**Comment by leotg130** [comment]
@gaearon Would also like to see this fixed. I think ErrorBoundary should have same semantics as exception handling, once an exception has been caught, it's caught.

If you want to log your exceptions, you should log them in the `catch()`, having a helper that logs the error exactly like it's currently logged might be useful. I guess this would be a breaking change, but IMO it's worth to have the same semantics as normal exception handling. If breaking change is too big of a problem, then having a `.preventDefault()`/`.preventLogging()` works too.

If the exception handler (ErrorBoundary) itself causes an exception, that's a different case, and should cause an exception to bubble up (which presumably could be caught by a higher level handler/ErrorBoundary).

For me there's two main reasons:
1) Nicer DX/UX, we know that in some cases some components will fail, and thus use fallback components, spamming the console isn't very nice. The console messages would lead people to think that something is broken, even though it has already been handled gracefully, by the fallback component.
2) E2e tests, we use playwright, and it would be nice to have all tests confirm at the end of the test that there were no log entries. It's too easy for developers to not notice something in the log, and having e2e tests that double check is useful.

**Comment by timredband** [comment]
My team ran into this issue as well. Would really like a fix.

**Comment by olee** [comment]
Yeah, being able to control this logging behavior would be awesome.
We are using an extensive logging system and it would be nice to be able to combine the error and the component stack into a single log message this way.
While just hiding the second component stack error message is not difficult, detecting the previous "uncaught" window error is the major issue as there's no ways to identify that it originated from within react.

**Comment by pleunv** [comment]
I'm exploring a pattern in which I throw specific errors when running into unexpected scenarios (i.e. missing or invalid parameters, invalid routes, ...) with the intention of moving all of this UI state logic to error boundaries instead rather than repeating it in every page/view. It makes things *a lot* cleaner, but the unfortunate side-effect is indeed that all of these errors now show up as uncaught errors in the console. Would really love a way to be able to bail out of logging expected errors.

**Comment by kevinhaube** [comment]
Error stacks are ludicrous when testing error boundaries 😬 yikes.

**Comment by epferrari** [comment]
bump

**Comment by samettttt** [comment]
bump

**Comment by AbhiPrasad** [comment]
The duplicate logging (once from console, once from error boundary), is troublesome for user's who've configured Sentry error monitoring that way.

Considering there exists API to do this for hydration errors with `onRecoverableError` https://github.com/facebook/react/pull/23207, is there more appetite for exposing this more publically?

I wouldn't even mind if it's a method I can monkeypatch with a Proxy, if labelled appropriately consumers can take the necessary risks. I did something similar with Next.js (https://github.com/vercel/next.js/discussions/36641) to log out hydration errors since Next.js framework doesn't expose this themselves.

**Comment by AlexGalays** [comment]
This is a huge pain and makes the console unreadable in dev mode as well as distracting in prod mode.

What I ended up doing to workaround this limitation:

1) Tag (e.g, `InvariantError extend Error`) my custom Errors
2) override console.error with my own to ignore any instance of the above custom error. This is for production builds.
3) Likewise, add an `error` listener on window and event.preventDefault() if event.error is an instance of my custom error. This is for dev builds. This is thanks to this amazing code: https://github.com/facebook/react/blob/main/packages/react-reconciler/src/ReactFiberErrorLogger.js#L93

It really shouldn't be this complicated and undocumented.

**Comment by staticshock** [comment]
I'm running into this as well for reasons similar to those explained by @AlexGalays. This is a rather unintuitive limitation.

**Comment by jfnault-seedbox** [comment]
bump

**Comment by durchanek** [comment]
Also spent several hours trying to figure out why there are still errors in the console after I had added an error boundary. But it looks like [React 19 adds options for handling errors](https://react.dev/blog/2024/04/25/react-19#error-handling) 🤞

**Comment by AbhiPrasad** [comment]
yes it's a good change - I think you can consider this issue closed, you can use the new options to override the default behaviour to ping `console.error` in React 19. Here's the setup I use in my test app:

```ts
import { createRoot } from "react-dom/client";

const container = document.getElementById(“app”);
const root = createRoot(container, {
  // Callback called when an error is thrown and not caught by an ErrorBoundary.
  onUncaughtError: (error, errorInfo) => {
    // only log in dev environments
    if (IS_DEV) {
      console.warn('Uncaught error', error, errorInfo.componentStack);
    }
  }),
  // Callback called when React catches an error in an ErrorBoundary.
  onCaughtError: (error, errorInfo) => {
    // do not log errors handled by error boundaries
  }),
  // Callback called when React automatically recovers from errors (for ex. hydration errors)
  onRecoverableError: (error, errorInfo) => {
    console.warn('Recoverable error', error, errorInfo.componentStack);
  }),
});

root.render();
```

**Comment by jtomaszewski** [comment]
Thanks @AbhiPrasad .

The default should be changed however. Caught errors shouldn't be logged to the console by default. That's how things work in general everywhere, e.g. in try-catch statements.
